### ISSUE TITLE
Nitpicky Door Defaults

### ISSUE THREAD

**Issue by TedCC1209** [CCDevNelson/Snap-Updates#63]
Can we have doors automatically come with a top KD? And if there is already a KD placed there (whether from the construction options or otherwise), have a rule so it doesn't export a duplicate KD?

Also, can we have doors in openings 21"-24" automatically come with the "Force Double Door" option selected? Then users can unselect the button if they want a single door in that section.

I think for right now it is fine to keep it impossible to have a single door on sections greater than 24" (right now doors placed in openings greater than 24" (can be 24.0000000001) automatically come with double doors, and because there is only the option to force a double door it is impossible to make it a single). I'm sure we will get some push back on this if we haven't already.

**Comment by CCDevNelson** [comment]
When we first developed this, I was told we don't always put a KD on the top of doors, and that is why you have the option to put a KD or a striker.

**Comment by TEDSTANDAGE** [comment]
It is rare that we use strikers. It should be default to have a top KD for the doors with an option to remove it, not the other way around

Further, my understanding is that the only way to put a top KD in is through the construction options in the partition prompts, not even in the door prompts, which isn’t ideal.

**Comment by TEDSTANDAGE** [comment]
To expound on this and clarify--"Base" and "Tall" doors currently default with a Top KD and not a Bottom KD, while "Upper" doors default with a Bottom KD and not a Top KD, perhaps expecting that the user would have put in the other KDs while setting up the partitions/openings. There is no option to add/delete a KD within the door prompts. I think the default should be that every door assembly entered comes with both Top and Bottom KDs with (1) a rule not to duplicate a KD if there is one already there from the partition/opening prompts or other assembly and (2) the option of deleting/adding the KDs (separate options for each KD) in the door prompts.

**Comment by riyajose96** [comment]
Added top/bottom KD and their prompts to door. Made corrections so as the KD moves to the expected location when "Fill Opening" is enabled. 
Can find the code at the branch -- 'riya-fixed'

**Comment by meltnaz** [comment]
Perhaps this is the issue that I'm seeing, but when I chose fill opening on the door prompt, it seems like door is 1 hole too long...maybe I'm seeing the KD in the wrong place? Just wanted to bring awareness of that issue.

**Comment by ClassyTeddy** [comment]
It seems at the moment, when the Fill Opening option is selected, the Top and Bottom KD's use their position before the fill opening option was selected. This will have to be fixed so that when the fill opening option is selected, the Top KD is placed at the Top of the partition, and the Bottom KD is placed at the bottom of the partition.

**Comment by TEDSTANDAGE** [comment]
Melissa, the issue of doors being one hole too big when using the fill opening feature is being addressed in SNaP-Dev issue number 167 and it is on Ryan’s to-do standard priority list.

**Comment by TEDSTANDAGE** [comment]
Teddy, perhaps the issues are related and you and Riya can attack both?

**Comment by ClassyTeddy** [comment]
It is very likely that Riya and I can take a look at it. So it is just one hole too long at the bottom?

**Comment by meltnaz** [comment]
I don't even know if it's a  full hole. Door bottom definitely goes past the top of the drawer face.
![door fill opening issue](https://user-images.githubusercontent.com/32275766/82264116-2871cc00-9919-11ea-885f-2377e720230a.png)

**Comment by TEDSTANDAGE** [comment]
The KDs are not being sized correctly and are not correctly adjusting the cleat locations properly. See the following example of an Upper Door Type with the Top and Bottom KD options selected on the door prompts (the top KD option is not selected in the partition prompts):

![image](https://user-images.githubusercontent.com/61025894/82455267-9b895880-9a67-11ea-87df-0cc15482cc80.png)

As you can see,  the top KD is going from the front of the partition to the top cleat (3/4" from the back of the partition) and the bottom KD is 3/4" set back from the depth of the partition where there is no cleat. When the Top KD is selected in the door prompts in this scenario, both KDs should extend the entire depth of the partition and the top cleat (and cover cleat) should be moved down 3/4" to account for the Top KD. You can see the top KD and adjustment of the cleat working properly when the top KD is selected in the partition prompts:

![image](https://user-images.githubusercontent.com/61025894/82455537-e30fe480-9a67-11ea-9625-18d447f90ac0.png)

This same thing is happening on Tall and Base door types just with the bottom cleat (and top cleat when "fill opening" is enable):

![image](https://user-images.githubusercontent.com/61025894/82456331-f53e5280-9a68-11ea-8f5e-0ff4de4b0803.png)

![image](https://user-images.githubusercontent.com/61025894/82456376-fff8e780-9a68-11ea-90b8-789f84e84c91.png)

Please see me if you have further questions.

**Comment by TEDSTANDAGE** [comment]
I'm not sure what you changed with the bottom KDs of Upper door types (or top KDs of Base/Tall door types) because they were sizing correctly before (and currently are in the master branch):

![image](https://user-images.githubusercontent.com/61025894/82457891-dc36a100-9a6a-11ea-8615-e4e0e6f2cc44.png)

But, please be careful in the changes you are making. Things may seem to be complicated in there now, but there is likely a reason. Please also keep in mind that the depth of the KDs will also need to adjust depending on the backing in the section, which is already currently functioning correctly in the master branch.

If this issue is getting to complicated we can shelve it for now since we can technically get everything in there correctly as it stands and these were just enhancement requests I was hoping we could do. Feel free to put this to the side and work on other issues.

**Comment by ClassyTeddy** [comment]
The placement/sizing of the top and bottom KD were adjusted to take into account the Top and bottom cleat. Moving the cleat up or down depending on if there is a Top or bottom KD selected in the door prompts will likely be a difficult fix for me or Riya, but doable. 

As for the KD already being placed properly in the master, keep in mind that the KD attached to the door in the master branch never has to take into account the cleats. When me and Riya were sizing, placing the KDs, we decided it was best to move them forward so they would not clip with the cleats.

**Comment by ClassyTeddy** [comment]
This new issue will likely be tackled when issue #65 is being worked on.

**Comment by TEDSTANDAGE** [comment]
Okay, well we shouldn't merge this fix with the master until these items are resolved. I suggest putting this to the side for now.

**Comment by ClassyTeddy** [comment]
All issues relating to this thread should be solved including:
* Doors automatically coming with Top and Bottom KDs (Available to test on `Teddy-Door-KD-Fixes`)
* Door openings between 21"and 24" automatically having force double door checked (Available to test on `riya-fixed`
*  Fill opening no longer makes the door too long (Available to test on `Teddy-Door-KD-Fixes`)
* Door KD's extend the entire depth of the partition and cleats move accordingly (Available to test on `Teddy-Door-KD-Fixes`)
* Door KD's now function dynamically when placed above a drawer stack, drawer stack **MUST** be placed first (Available to test on `Teddy-Door-KD-Fixes`)

Because all of these issues have been addressed, I am going to close this issue.

**Comment by ClassyTeddy** [comment]
Which need addressed specifically?

**Comment by TedCC1209** [comment]
Certainly:

-Doors automatically coming with Top and Bottom KDs (Available to test on Teddy-Door-KD-Fixes)
-Door openings between 21"and 24" automatically having force double door checked (Available to test on riya-fixed)
-Door KD's extend the entire depth of the partition and cleats move accordingly (Available to test on Teddy-Door-KD-Fixes)--This one looks like the top KD correctly extends the entire depth but the top cleat does not move down when the top KD is present (the top cleat moves down if you select Top KD in the construction options for the partitions, but not if you select Top KD in the doors prompts--again though, there is no "Top KD" option in the door prompts if the door is an upper, only if it is a tall or base cabinet)

The fill opening stuff still needs some help (especially with backs and as they relate to other assemblies within the opening), but that can be addressed later and separately.

**Comment by ClassyTeddy** [comment]
So these two work for me on the master branch:
-Doors automatically coming with Top and Bottom KDs
-Door KD's extend the entire depth of the partition and cleats move accordingly

I will look into the automatically having force double door checked

**Comment by TedCC1209** [comment]
> So these two work for me on the master branch:
> -Doors automatically coming with Top and Bottom KDs
> -Door KD's extend the entire depth of the partition and cleats move accordingly

These work when there is nothing in the opening prior to placing the doors. However, these act like I wrote above and don't work properly when placed above drawers and hampers

**Comment by ClassyTeddy** [comment]
Oh, they never functioned that way. That will require far more logic. I can try and look into a solution

**Comment by ClassyTeddy** [comment]
There is a possible solution for all the issues brought up here on the `Door-Defaults-Working-Above-Drawers` branch. This fix should work for both drawers and hampers. This branch also fixed the automatically having double doors be selected, but this only works on placement of the doors, as it was working before. This branch also fixes issue #245, as there were some duplications and incorrect indentations I noticed when working on this issue that I fixed. Please do rigorous testing with doors on this branch, as the changes made affect a lot of logic, including that of the fill opening option, the top and bottom KD's, and the logic for BCP's.

**Comment by TedCC1209** [comment]
> There is a possible solution for all the issues brought up here on the `Door-Defaults-Working-Above-Drawers` branch. This fix should work for both drawers and hampers. This branch also fixed the automatically having double doors be selected, but this only works on placement of the doors, as it was working before. This branch also fixes issue #245, as there were some duplications and incorrect indentations I noticed when working on this issue that I fixed. Please do rigorous testing with doors on this branch, as the changes made affect a lot of logic, including that of the fill opening option, the top and bottom KD's, and the logic for BCP's.

@sarahhightower @kategodfrey @JBTattersall  Just making sure you guys are seeing this

**Comment by kategodfrey** [comment]
I tested: Door-Defaults-Working-Above-Drawers and it appears to have the fixes for the automatic KD placements and the other door issues from 245. One thing I noticed is that if you delete the doors the KDs stay @TedCC1209 but maybe that's not a big deal.

**Comment by sarahhightower** [comment]
Automatic KD placements and door/shelf prompts appear to be working fine. In one of my test runs, however; I had an issue getting the door prompt to open. I'm attaching a visual for this. Let me know if you would like me to post the ccp file or if this is enough to understand the issue. 

![Uploading SNaP_ [C__Users_shightower_Documents_SNaP Projects_Test Room_41110_Testing Doors.blend] 12_21_2020 3_55_13 PM.png…]()

**Comment by TedCC1209** [comment]
Teddy, with BCPs the doors aren't reading the opening correctly so it is automatically putting in double doors in small spaces--can you have it figure out the opening space (cab opening width - BCP depth + 0.75" outside partition overlap) when determining whether to have Force Double Doors selected? Right now it is just looking at cab openings width I believe.

I could not replicate Sarah's door prompt error.

Otherwise this looks good to me. The KDs staying after deleting a door assembly is okay with me for now. Thanks, Ted

**Comment by TedCC1209** [comment]
Here's a pic of the BCP small double doors:

![small double doors](https://user-images.githubusercontent.com/61025894/102893264-853ce400-441e-11eb-8378-01aca38b24fe.png)

**Comment by ClassyTeddy** [comment]
The BCP Small double doors issue is fixed and ready for testing on the `Door-Defaults-Working-Above-Drawers` branch.

**Comment by TedCC1209** [comment]
Looks good, thanks Teddy

**Comment by TedCC1209** [comment]
@ClassyTeddy I think this is probably good to go. Can you do a pull request on this? These door items should likely be fixed in the master before I prep a new hinge drilling test (not that it necessarily affects the drilling, but just in case).

**Comment by kategodfrey** [comment]
Tested the BCP. Looks good.
